### ISSUE TITLE
Crash on ios12 beta with iphone7

### ISSUE THREAD

**Issue by arbalest876** [superk589/DereGuide#24]
Crash on ios12 beta with iphone7

**Comment by superk589** [comment]
```
The entity 'Unit' is not visible to this managed object context.
(
	0   CoreFoundation                      0x00000001831bf048 <redacted> + 176
	1   libobjc.A.dylib                     0x00000001824784e8 objc_exception_throw + 56
	2   CoreData                            0x0000000185ce6480 <redacted> + 916
	3   libswiftCoreData.dylib              0x0000000106bfa690 $SSo22NSManagedObjectContextC8CoreDataE5fetchySayxGSo14NSFetchRequestCyxGKSo0gH6ResultRzlF + 72
	4   DereGuide                           0x0000000104984850 $S9DereGuide15SyncCoordinatorC26fetchLocallyTrackedObjects33_93B7439E7235CA88941701F093A078ACLLyyFyycfU_yycfU_yycfU_yycfU_yycfU_ + 212
	5   DereGuide                           0x0000000104941528 $SIeg_IeyB_TR + 52
	6   SwiftTryCatch                       0x0000000105961948 +[SwiftTryCatch try:catch:finally:] + 136
	7   DereGuide                           0x0000000104984754 $S9DereGuide15SyncCoordinatorC26fetchLocallyTrackedObjects33_93B7439E7235CA88941701F093A078ACLLyyFyycfU_yycfU_yycfU_yycfU_ + 508
	8   DereGuide                           0x0000000104941528 $SIeg_IeyB_TR + 52
	9   CoreData                            0x0000000185dab854 <redacted> + 156
	10  libdispatch.dylib                   0x00000001071fd2d0 _dispatch_client_callout + 16
	11  libdispatch.dylib                   0x0000000107201fc0 _dispatch_main_queue_callback_4CF + 1360
	12  CoreFoundation                      0x00000001831714fc <redacted> + 12
	13  CoreFoundation                      0x000000018316f1b8 <redacted> + 1972
	14  CoreFoundation                      0x00000001830a8abc CFRunLoopRunSpecific + 436
	15  GraphicsServices                    0x00000001853ca578 GSEventRunModal + 100
	16  UIKitCore                           0x00000001ae5247b8 UIApplicationMain + 212
	17  DereGuide                           0x0000000104ceb310 main + 88
	18  libdyld.dylib                       0x0000000182c224d0 <redacted> + 4
)
```
I'm not sure why this exception occurs when call context.fetch(request). Everything was right before iOS 12.

**Comment by superk589** [comment]
It's fixed in iOS 12 beta 2.

**Comment by superk589** [comment]
Though it does not crash on iOS 12 beta 2, there are new error messages.

CoreData: error: The fetch request's entity 0x2804d35a0 'Unit' appears to be from a different NSManagedObjectModel than this context's

**Comment by DJ-Glock** [comment]
@superk589 
Hello
Have you find the way to fix the following error on iOS 12?

> CoreData: error: The fetch request's entity 0x2804d35a0 'Unit' appears to be from a different NSManagedObjectModel than this context's

**Comment by superk589** [comment]
@DJ-Glock 
Yes, this error is caused by that the entity (instance of NSEntityDescription) is used in a different managed context.
However, before iOS 12 beta, this is not an issue and will not cause a crash.  
The error code uses this code to initialize NSEntityDescription:
```
let entity = NSEntityDescription()
// "Unit" is the name of a custom entity.
entity.name = "Unit"
entity.managedObjectClassName = "Unit"
```

I fixed it by using another function only available in iOS 10.0+, it's a class function of NSManagedObject
```
class func entity() -> NSEntityDescription
```
The entity it returned always uses the same context as the object.

Finally, the code is like this:
```
extension NSEntityDescription {
    public convenience init<A: NSManagedObject>(cls: A.Type, name: String) {
        self.init()
        let a = NSStringFromClass(cls) as String
        self.managedObjectClassName = a
        self.name = name
    }
}

extension Managed where Self: NSManagedObject {
     public static var entity: NSEntityDescription {
         if #available(iOS 10.0, *) {
             return entity()
         } else {
             return NSEntityDescription(cls: Self.self, name: entityName)
         }
     }
}

**Comment by DJ-Glock** [comment]
@superk589  Thanks for your explanation. I will try to understand how to apply your solution to mine.
